architecture-guide.md

```diff
@@ -0,0 +1,425 @@
+
+  When stores are singleton, UI components can access backend from arbitrary
+  places, in an implicit way. It makes dependencies hard to track and APIs are not
+  clear. Singletons make the application hard to test (singleton mocking is problematic).
+  Also, the initialization order of singletons may be a problem in some cases.
+
+  ```qml
+  isGoerliEnabled: root.rootStore.isGoerliEnabled         // OK
+  isGoerliEnabled: WalletStores.RootStore.isGoerliEnabled // Bad
+  ```
+
+- Stores are a thin wrapper over context properties exposed from the backend, no
+  additional logic should be there like e.g. data transformations done by proxy models.
+
+  Stores are cut-off line in tests and `Storybook pages`. When additional logic
+  is put there, it's not available in unit tests and `Storybook`. It leads to
+  more complicated mocking and code duplication (same logic repeated in real
+  stores and mocked ones).
+
+  ```qml
+  // WalletAssetsStore.qml
+  
+  QtObject {
+      property LeftJoinModel groupedAccountAssetsModel: LeftJoinModel {
+          leftModel: root.baseGroupedAccountAssetModel
+          rightModel: _jointTokensBySymbolModel
+          joinRole: "tokensKey"
+      } // Bad - transformation should be kept in adaptor object,
+        // only basic models should be exposed from store
+  }
+  ```
+
+- Stores completely hide the backend's context properties and not expose them
+  directly.
+
+  It makes the contract between backend and frontend clear already from the
+  perspective of the store. It should be enough for a UI developer to rely on
+  the store's API without inspection of the nim code in order to discover what
+  methods and properties are exposed from objects directly exposed to the UI
+  from the backend via store.
+
+  ```qml
+  // WalletStore.qml
+  
+  QtObject {
+      property var networksModuleInst: networksModule // BAD
+  }
+  ```
+  
+  ```qml
+  // WalletStore.qml
+  
+  QtObject {
+      id: root
+  
+      readonly property bool isGoerliEnabled: networksModule.isGoerliEnabled
+  
+      signal chainIdFetchedForUrl(string url, int chainId)
+  
+      function foo() {
+          return networksModule.foo()
+      }
+  
+      Component.onCompleted: {
+         networksModule.chainIdFetchedForUrl.connect(root.chainIdFetchedForUrl)
+      }
+  }
+  ```
+
+  This additional layer causes that backend changes to affect only this part of
+  the UI code, which can be adjusted here even without changing the API of the
+  store itself. It's also clear what exactly is exposed.
+
+- A single context property injected from the backend should be exposed only
+  once in a single store.
+
+  Bad:
+
+  ```qml
+  // SwapStore.qml
+  QtObject {
+    readonly property var flatNetworks: networksModule.flatNetworks
+  }
+
+  // CommunityTokensStore.qml
+  QtObject {
+    readonly property var flatNetworks: networksModule.flatNetworks
+  }
+  ```
+
+  Stores represent the state of the application but they are not aware how that
```

Comment:
Maybe we could add here that a `store` should represent a specific `context` of the application and not more?!